sys/src/9/port/devramfs.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * This file is part of the UCB release of Plan 9. It is subject to the license
+ * terms in the LICENSE file found in the top-level directory of this
+ * distribution and at http://akaros.cs.berkeley.edu/files/Plan9License. No
+ * part of the UCB release of Plan 9, including this file, may be copied,
+ * modified, propagated, or distributed except according to the terms contained
+ * in the LICENSE file.
+ */
+
+#include	"u.h"
+#include	"../port/lib.h"
+#include	"mem.h"
+#include	"dat.h"
+#include	"fns.h"
+#include	"../port/error.h"
+
+#define DIR_EXTEND_SIZE 10
+
+int memcpy(void* dst, const void* src, int len);
+
+struct Directory {
+	int inuse;
+	int allocated;
+	Dirtab ramdir[];
+};
+
+struct RamFile {
+	int length;
+	char data[];
+};
+
+static struct Directory *ramroot;
+
+static char* readme = "This is the kernel ramdisk.\n";
+
+static void
+raminit(void)
```

Comment:
In general this is a non-plan9 way to do this. It's a familiar way esp. if you're used to linux. But the more idiomatic way is to use a gen function; see devproc.c for an example. 

In general plan 9 drivers don't mirror a file system structure in the data structures in the driver; rather, they respond to walks and such using a gen function. It's a bit of a subtle distinction but important. You might want to look at devproc.c to see how it's done.  devproc.c is particularly interesting because procs are so dynamic, and the driver has to deal with that, even during directory reads. it also has the locking right.